src/jesse_state.erl

```diff
@@ -182,7 +182,11 @@ set_current_schema(#state{id = Id} = State, NewSchema0) ->
         %% Instead of just removing all the other fields, we put schema as
         %% 1st element, so, only `ref' will be validated, while other fields
         %% (say, `definitions') may still be referenced.
-        [{?REF, Ref} | lists:keydelete(?REF, 1, NewSchema0)]
+        ListSchema = case is_map(NewSchema0) of
+                       true -> maps:to_list(NewSchema0);
```

Comment:
Why convert map to list? It seems like map is a superior data storage mechanism for performance. It also seems like you don't need to do anything at all here for the map case (map keys are unique, unordered). Maybe I don't understand the issue that is solved here very well.

Reply:
The thig is that we would alwasy convert schema to list anyway before applying it:

https://github.com/for-GET/jesse/blob/6cfd1c09ff7abfcbce445e44583cd3a41a5935af/src/jesse_schema_validator.erl#L89
https://github.com/for-GET/jesse/blob/6cfd1c09ff7abfcbce445e44583cd3a41a5935af/src/jesse_json_path.erl#L125-L134

We could, in theory, try to use `maps:iterator()` to iterate over map-schema, but that would require quite a lot of work with no clear benefit (because rules in schema are applied iteratively, one-by-one).

Reply:
And the issue this code solves is:

When schema contains a reference AND some other validators, and also some definitions to be referenced, say:

```json
{
 "type": "object",
 "$ref": "#/path/to/other/schema",
 "definitions": {
   "a": {...}
 }
}
````

(which is kind of artificial, but it is included in the official JSON schema test suite), the following conditions should be true:

* any other validators should be ignored, when there is a `$ref` in a schema, so, `"type": "object"` rule should not be applied
* schema from the reference should "replace" the current schema
* but all the keys defined in the same object as the one containing '$ref' are still should be accessible for referencing, so, it still should be possible to reference `definitions.a` from somewhere

Which means we should at the same time ignore all the other validators in current schema AND we can't just remove all the other keys (that's why previous attempt to solve this by just removing all the other keys from object containing `$ref` failed after we updated the test suite)